Add numeral system translator helper

jamesdonoh opened the pull request:
Resolves https://github.com/bbc/psammead/issues/2299

**Overall change:** _Adds a utility for converting numbers to an arbitrary specified numeral system_

**Code changes:**

- Added `makeNumeralTranslator` helper function to `/numerals`
- Added unit test coverage `makeNumeralTranslator`
- Updated README with description and example usage

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated jest tests added (for new features) or updated (for existing features)
- [NA] This PR requires manual testing

jamesdonoh committed:
Add initial unit tests for numeral translator

jamesdonoh committed:
Update snapshots following test restructure

jamesdonoh committed:
Add initial implementation for makeNumeralTranslator

jamesdonoh committed:
Document makeNumeralTranslator helper

jamesdonoh committed:
Split validation into helper for readability

jamesdonoh committed:
Clarify documentation for makeNumeralTranslator

jamesdonoh committed:
More documentation tweaks for makeNumeralTranslator

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into numeral-translator

jamesdonoh committed:
Bump version to 2.23.0 for makeNumeralTranslator

rhenshaw56 commented on packages/utilities/psammead-locales/src/numerals/__snapshots__/index.test.js.snap:
```diff
@@ -63,3 +63,5 @@ Array [
 ]
 `;
+
+exports[`Numeral systems should return makeNumeralTranslator numerals 1`] = `[Function]`;
```
`makeNumeralTranslator` doesn't have numerals, I don't think it should be in this snap

jamesdonoh replied:
True, it's not used, but `jest -u` adds it automatically, not sure how to suppress it?

rhenshaw56 replied:
@jamesdonoh I think it has to do with how it's been imported
```
import * as numerals from '.'; // will import all exports in '.'
```

simonsinclair replied:
We could suppress this by being more explicit about what's imported. https://gist.github.com/simonsinclair/ccc86b2ff9d900e3a08f1dd4be491fcb#file-index-test-js-L1-L10

jamesdonoh replied:
Oh yes, good idea.

jamesdonoh replied:
I've gone with @rhenshaw56's `Array` check actually as it reduces the need to keep updating the `import` list every time we add a new numeral system.

rhenshaw56 commented on packages/utilities/psammead-locales/src/numerals/index.test.js:
```diff
@@ -1,10 +1,46 @@
+    });
+  });
+
+  it('invalid numeral systems', () => {
```
maybe this should be a `describe` block having
```
it.each(invalidSystems)('should return ...
```
or still an `it` block with `invalid numeral systems should return something`

jamesdonoh replied:
I did this originally but I thought it was unhelpful to list all the invalid systems in the test output so I removed it.

jamesdonoh replied:
I've wrapped it in a describe block as you suggest so the output is a bit more descriptive, now it's:

```
  makeNumeralTranslator
    invalid numeral systems
      ✓ should return the input unchanged
```

Please take a look.

jamesdonoh committed:
Add describe block to clarify test structure following CR

rhenshaw56 committed:
Merge branch 'latest' into numeral-translator

simonsinclair reviewed (commented):
This is so simple - didn't take me long to understand. 👍 All tests pass.
It would be cool if we could verify the order of a `system`'s numerals, but they don't change that often and are covered by the snapshots.

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into numeral-translator

simonsinclair committed:
Merge branch 'latest' into numeral-translator

jamesdonoh committed:
Exclude makeNumeralTranslator from snapshot tests

jamesdonoh committed:
Merge branch 'numeral-translator' of github.com:bbc/psammead into numeral-translator

jamesdonoh committed:
Exclude makeNumeralTranslator from Storybook

rhenshaw56 reviewed (approved):
Looks good.

rhenshaw56 committed:
Merge branch 'latest' into numeral-translator

jamesdonoh committed:
Support numeral systems with more than 10 symbols

Currently this only uses the first 10 symbols when translating strings,
so numeral systems which support a different symbol for 10 may not work
as expected.

jamesdonoh committed:
Merge branch 'numeral-translator' of github.com:bbc/psammead into numeral-translator

jamesdonoh committed:
Merge branch 'latest' of github.com:bbc/psammead into numeral-translator

jamesdonoh commented:
@HarveyPeachey pointed out a problem with this, which is that the preexisting numeral systems exported from `/numerals` like `EasternArabic` actually all have 11 symbols not the 10 expected by this helper.

To address this I've pushed a fix https://github.com/bbc/psammead/pull/2300/commits/28341ceb7da2e5861f85b23ea882a5e72366a0c2 so that these systems are accepted, but that the translator only uses the first 10 symbols of any system.
